File: orion-viewer/src/universe/constellation/orion/viewer/selection/SelectionAutomata.java
/*
 * Orion Viewer - pdf, djvu, xps and cbz file viewer for android devices
 *
 * Copyright (C) 2011-2013  Michael Bogdanov & Co
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package universe.constellation.orion.viewer.selection;

import android.R;
import android.app.Dialog;
import android.graphics.Rect;
import android.view.*;
import universe.constellation.orion.viewer.OrionView;
import universe.constellation.orion.viewer.OrionViewerActivity;

/**
 * User: mike
 * Date: 11.08.12
 * Time: 14:37
 */
public class SelectionAutomata {

    private enum STATE {START, MOVING, END, CANCELED};

    private STATE state = STATE.CANCELED;

    private int startX, startY, width, height;

    private Dialog selectionDialog;

    private SelectionView selectionView;

    private OrionViewerActivity activity;

    public SelectionAutomata(final OrionViewerActivity activity) {
        this.activity = activity;

        selectionDialog = new Dialog(activity, R.style.Theme_Translucent_NoTitleBar);
        selectionDialog.setContentView(universe.constellation.orion.viewer.R.layout.text_selector);

        View view = activity.getLayoutInflater().inflate(universe.constellation.orion.viewer.R.layout.text_selector, null);

        selectionDialog.setContentView(view);



        selectionView = (SelectionView) selectionDialog.findViewById(universe.constellation.orion.viewer.R.id.text_selector);

        selectionView.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View v, MotionEvent event) {
                return SelectionAutomata.this.onTouch(event);
            }
        });
    }



    public boolean onTouch(MotionEvent event) {
        int action = event.getAction();

        //System.out.println("aaaction " + action + " " + event.getX() + ", " + event.getY());
        STATE oldState = state;
        boolean result = true;
        switch (state) {
            case START:
                if (action == MotionEvent.ACTION_DOWN) {
                    startX = (int) event.getX();
                    startY = (int) event.getY();
                    width = 0;
                    height = 0;
                    state = STATE.MOVING;
                    selectionView.reset();
                } else {
                    state = STATE.CANCELED;
                }
                break;

            case MOVING:
                int endX = (int) event.getX();
                int endY = (int) event.getY();
                width = endX - startX;
                height = endY - startY;
                if (action == MotionEvent.ACTION_UP) {
                    state = STATE.END;
                } else {
                    selectionView.updateView(Math.min(startX, endX), Math.min(startY, endY), Math.max(startX, endX), Math.max(startY, endY));
                }
                break;

            default: result = false;
        }

        if (oldState != state) {
            switch (state) {
                case CANCELED: selectionDialog.dismiss(); break;

                case END:
                    selectionDialog.dismiss();

                    String text = activity.getController().selectText(getStartX(), getStartY(), getWidth(), getHeight());
                    if (text != null && !"".equals(text)) {
                        new SelectedTextActions(activity).show(text);
                    } else {
                        activity.showFastMessage(universe.constellation.orion.viewer.R.string.warn_no_text_in_selection);
                    }
                    break;
            }
        }
        return result;
    }

    public void startSelection() {
        selectionView.reset();
        initDialogSize();
        selectionDialog.show();

        activity.showFastMessage("Select text!");
        state = STATE.START;
    }

    private void initDialogSize() {
        OrionView orionView = activity.getView();
        Rect rect = orionView.getViewCoords();
        int[] coords = new int[]{rect.left, rect.top};
        orionView.getLocationOnScreen(coords);
        int left = coords[0];
        int top = coords[1];
        int width = rect.width();
        int height = rect.height();
        WindowManager.LayoutParams params = selectionDialog.getWindow().getAttributes();
        params.width = width;
        params.height = height;
        params.x = left;
        params.y = top;
        selectionDialog.getWindow().setAttributes(params);
    }

    public boolean inSelection() {
        return state == STATE.START || state == STATE.MOVING;
    }

    public boolean isSuccessful() {
        return state == STATE.END;
    }

    public int getStartX() {
        return startX;
    }

    public int getStartY() {
        return startY;
    }

    public int getWidth() {
        return width;
    }

    public int getHeight() {
        return height;
    }

}

File: orion-viewer/src/main/java/universe/constellation/orion/viewer/selection/SelectionAutomata.kt
package universe.constellation.orion.viewer.selection

import android.content.DialogInterface
import android.graphics.PointF
import android.graphics.RectF
import android.view.MotionEvent
import android.view.View
import androidx.core.graphics.toRect
import universe.constellation.orion.viewer.Action
import universe.constellation.orion.viewer.Controller
import universe.constellation.orion.viewer.OrionViewerActivity
import universe.constellation.orion.viewer.R
import universe.constellation.orion.viewer.dialog.DialogOverView
import universe.constellation.orion.viewer.document.TextAndSelection
import universe.constellation.orion.viewer.dpToPixels
import universe.constellation.orion.viewer.log
import universe.constellation.orion.viewer.view.PageLayoutManager
import kotlin.math.max
import kotlin.math.min

class SelectionAutomata(val activity: OrionViewerActivity) :
    DialogOverView(activity, R.layout.text_selector, android.R.style.Theme_Translucent_NoTitleBar) {
    private enum class STATE {
        START, ACTIVE_SELECTION, MOVING_HANDLER, CANCELED
    }

    private var state = STATE.CANCELED

    private val selectionView: SelectionViewNew = dialog.findViewById(R.id.text_selector)

    private var isSingleWord = false

    private var translate = false

    private val triangleSide: Float

    private val singleWordDelta: Float

    private var activeHandler: Handler? = null

    private var actions: SelectedTextActions? = null

    private var lastX: Float = 0f
    private var lastY: Float = 0f

    private var isRectSelection = true

    init {
        selectionView.setOnTouchListener { v: View?, event: MotionEvent ->
            this@SelectionAutomata.processTouch(
                event
            )
        }
        triangleSide = activity.dpToPixels(30f).toFloat()
        singleWordDelta = activity.dpToPixels(2f).toFloat()
    }

    private fun debug(msg: String) {
        log("touch: $msg")
    }

    private fun processTouch(event: MotionEvent): Boolean {
        val action = event.action
        debug("action: $action oldState: $state")

        val oldState = state
        var result = true
        when (state) {
            STATE.START -> if (action == MotionEvent.ACTION_DOWN) {
                selectionView.reset()
                setStartHandlers(event.x, event.y)

                activeHandler = selectionView.endHandler
                if (!isSingleWord) {
                    state = STATE.MOVING_HANDLER
                }
            } else {
                state = if (isSingleWord && action == MotionEvent.ACTION_UP) {
                    STATE.ACTIVE_SELECTION
                } else if (isSingleWord && action == MotionEvent.ACTION_MOVE) {
                    STATE.START
                } else {
                    STATE.CANCELED
                }
            }

            STATE.ACTIVE_SELECTION -> {
                if (action == MotionEvent.ACTION_DOWN) {
                    activeHandler =
                        selectionView.findClosestHandler(event.x, event.y, 0.8f * triangleSide)
                    println(activeHandler)
                    if (activeHandler == null) {
                        state = STATE.CANCELED
                        result = false
                    } else {
                        state = STATE.MOVING_HANDLER
                        isSingleWord = false
                        actions?.dismissOnlyDialog()
                    }
                } else {
                    //cancel
                }
            }

            STATE.MOVING_HANDLER -> if (action == MotionEvent.ACTION_UP || action == MotionEvent.ACTION_MOVE) {
                activeHandler!!.x += event.x - lastX
                activeHandler!!.y += event.y - lastY
                recordLastEvent(event)

                updateSelectionAndGetText(getSelectedText())

                if (action == MotionEvent.ACTION_UP) {
                    state = STATE.ACTIVE_SELECTION
                }
            } else {
                result = false
            }

            else -> result = false
        }

        if (oldState != state) {
            debug("new state: $action oldState: $state")
            when (state) {
                STATE.CANCELED -> {
                    actions?.dismissOnlyDialog()
                    actions = null
                    dialog.dismiss()
                }

                STATE.ACTIVE_SELECTION -> {
                    val text = getSelectedText()
                    if (oldState == STATE.START && isSingleWord) {
                        debug("oldState == STATE.START && isSingleWord")
                        text?.let {
                            it.rect.firstOrNull()?.let {
                                debug("rect: $it")
                                updateHandlersByTextSelection(it)
                            }
                        }
                    }
                    updateSelectionAndGetText(text)
                    showActionsPopupOrDoTranslation(
                        text?.value ?: "",
                        translate,
                        activity.controller!!
                    )
                }

                STATE.MOVING_HANDLER -> {
                    recordLastEvent(event)
                }

                else -> {
                }
            }
        }
        debug("result = $result")
        return result
    }

    private fun recordLastEvent(event: MotionEvent) {
        lastX = event.x
        lastY = event.y
    }

    private fun setHandlers(start: PointF, end: PointF) {
        selectionView.setHandlers(
            Handler(
                start.x,
                start.y,
                triangleSide,
                true
            ), Handler(
                end.x,
                end.y,
                triangleSide,
                false
            )
        )
    }

    private fun updateSelectionAndGetText(text: TextAndSelection?) {
        selectionView.updateView(text?.rect ?: emptyList())
        selectionView.invalidate()
    }

    private fun getSelectedText(): TextAndSelection? {
        return getTextByHandlers(
            selectionView.startHandler!!,
            selectionView.endHandler!!,
            isRect = isRectSelection,
            isSingleWord = isSingleWord
        )
    }

    fun initSelectionByPosition(pos: ClickInfo, translate: Boolean) {
        resetState(true, translate)
        isRectSelection = false

        setStartHandlers(pos.x.toFloat(), pos.y.toFloat())

        val selectedText = getSelectedText()
        val controller = activity.controller ?: return

        selectedText?.rect?.let {
            it.firstOrNull()?.let { rect ->
                updateHandlersByTextSelection(rect)
            }
        }

        updateSelectionAndGetText(selectedText)
        showActionsPopupOrDoTranslation(selectedText?.value ?: "", translate, controller)
    }

    private fun setStartHandlers(x: Float, y: Float) {
        val selection = getScreenSelectionRectWithDelta(
            RectF(
                x,
                y,
                x,
                y,
            ), isSingleWord
        )
        setHandlers(
            PointF(selection.left, selection.top),
            PointF(selection.right, selection.bottom)
        )
    }

    private fun updateHandlersByTextSelection(rect: RectF) {
        selectionView.startHandler!!.apply {
            x = rect.left
            y = rect.top
        }
        selectionView.endHandler!!.apply {
            x = rect.right
            y = rect.bottom
        }
    }

    private fun showActionsPopupOrDoTranslation(
        text: String,
        translate: Boolean,
        controller: Controller
    ) {
        if (translate && text.isNotBlank()) {
            dialog.dismiss()
            Action.DICTIONARY.doAction(controller, activity, text)
        } else {
            val selectedTextActions = SelectedTextActions(activity, dialog)
            actions = selectedTextActions
            val occupiedArea = getScreenSelectionRect(selectionView.startHandler!!, selectionView.endHandler!!)
            occupiedArea.inset(0f, -triangleSide)
            if (!dialog.isShowing) {
                //TODO: refactor
                dialog.setOnShowListener { dialog2: DialogInterface? ->
                    selectedTextActions.show(text, occupiedArea)
                    dialog.setOnShowListener(null)
                }
                startTextSelection(true, translate = false, quite = true)
                state = STATE.ACTIVE_SELECTION
            } else {
                selectedTextActions.show(text, occupiedArea)
            }
        }
    }

    @JvmOverloads
    fun startTextSelection(isSingleWord: Boolean, translate: Boolean, quite: Boolean = false) {
        selectionView.setColorFilter(activity.fullScene.colorStuff.backgroundPaint.colorFilter)
        if (!quite) {
            selectionView.reset()
        }
        initDialogSize()
        dialog.show()
        if (!quite) {
            val msg =
                activity.resources.getString(if (isSingleWord) R.string.msg_select_word else R.string.msg_select_text)
            activity.showFastMessage(msg)
            resetState(isSingleWord, translate)
            isRectSelection = !isSingleWord
        }
    }

    private fun resetState(isSingleWord: Boolean, translate: Boolean) {
        state = STATE.START
        this.isSingleWord = isSingleWord
        this.translate = translate
        this.isRectSelection = true
        actions?.dismissOnlyDialog()
        actions = null
        activeHandler = null
    }

    fun getPageSelectionRectangles(
        rect: RectF,
        isSingleWord: Boolean,
        pageLayoutManager: PageLayoutManager
    ): List<PageAndSelection> {
        val expandedRect = getScreenSelectionRectWithDelta(rect, isSingleWord)
        return pageLayoutManager.findPageAndPageRect(expandedRect.toRect())
    }

    private fun getScreenSelectionRectWithDelta(
        rect: RectF,
        isSingleWord: Boolean
    ): RectF {
        if (!isSingleWord) return rect
        rect.inset(-singleWordDelta, -singleWordDelta) //TODO: dp to pixel
        return rect
    }

    companion object {
        fun getScreenSelectionRect(startHandler: Handler, endHandler: Handler): RectF {
            return RectF(
                min(startHandler.x, endHandler.x),
                min(startHandler.y, endHandler.y),
                max(startHandler.x, endHandler.x),
                max(startHandler.y, endHandler.y)
            )
        }
    }
}
